design/runnable.go

```diff
@@ -78,125 +96,581 @@ var _ = Service("runnable", func() {
+			Example("/workspace/input/configuration.txt")
+		})
+	Field(5, "labels", MapOf(String, String),
+		"List of custom labels used to determine how to connect this input parameter to outputs of other runnables.",
+		func() {
+			Key(func() {
+				Pattern(identifierPattern)
+			})
+			Example(map[string]string{
+				"library":  "pytorch",
+				"function": "predict",
+			})
+		})
 })
 
-// InputParameter describes runnable input parameter
-var InputParameter = Type("InputParameter", func() {
-	Field(1, "datatype", String, "Parameter data type", func() {
-		Example("file")
+// RunnableOutputParameter describes a runnable output parameter
+var RunnableOutputParameter = Type("RunnableOutputParameter", func() {
+	Field(1, "description", String, "Parameter description", func() {
+		MaxLength(1000)
+		Default("")
+		Example("Output parameter description")
 	})
-	Field(2, "optional", Boolean, "Optional parameter", func() {
-		Example(true)
-	})
-	Field(3, "default", String, "Default value", func() {
-		Example("mydata.csv")
+	Field(2, "optional", Boolean, "Optional output parameter", func() {
+		Default(false)
 	})
+	Field(3, "defaultValue", String, "Default value for optional output parameters")
+	Field(4, "path", String,
+		"Specify a custom container path where the output parameter value is provided by the container as a file",
+		func() {
+			Example("/workspace/output/model-url.txt")
+		})
+	Field(5, "labels", MapOf(String, String),
+		"List of custom labels used to determine how to connect this input parameter to outputs of other runnables.",
+		func() {
+			Key(func() {
+				Pattern(identifierPattern)
+			})
+			Example(map[string]string{
+				"library":  "pytorch",
+				"function": "predict",
+			})
+		})
 })
 
-// RunnabelOutput describes output of a runnable
-var RunnabelOutput = Type("RunnableOutput", func() {
-	Field(1, "name", String, "Output name", func() {
-		Example("Output1")
+// RunnableInputArtifact describes a runnable input artifact
+var RunnableInputArtifact = Type("RunnableInputArtifact", func() {
+	Field(1, "description", String, "Artifact description", func() {
+		MaxLength(1000)
+		Default("")
+		Example("Artifact description")
 	})
-	Field(2, "kind", String, "Kind of output (e.g. runnable, dataset, model, metatada, etc.)", func() {
-		Example("model")
+	Field(2, "optional", Boolean, "Optional input artifact", func() {
+		Default(false)
 	})
-	Field(3, "runnable", RunnableRef, "Runnable reference")
-	Field(4, "metadata", InputParameter, "Metadata description")
+	Field(3, "provider", ArrayOf(String),
+		`Data passing mechanisms supported by the runnable implementation used to provide
+the artifact's contents to the container. The order is significant: the framework will choose the
+first mechanism in the supplied list that matches the concrete input artifact supplied at
+runtime as well as other conditions such as the workflow where this runnable is referenced,
+the way its inputs and outputs are connected to other runnables etc. 
```

Comment:
Could you add an example of how the order is significant?

Reply:
This depends on how we reference a workflow template to a runnable, more specifically how a workflow run created from a workflow template that references a runnable provides the concrete input artifacts. For example, let's assume a runnable has a codeset input artifact listing both `git` and `local` as providers (in this order). If at runtime the input value is provided as an codeset reference, the `git` provider is chosen and the codeset is passed as a git URL to the container. If at runtime, the input value is connected to the codeset output of some other runnable also using the `local` provider, the `local` provider is chosen and the codeset is passed to the container using a local container path (i.e. the workspace path where the other runnable saved the codeset contents).

It's difficult to explain in this help text how that would work, and it won't be clarified until we actually implement the workflow template feature, but so much is clear: the `provider` field can be used to "tell" the framework which mechanisms/protocols/services the container supports for uploading/downloading artifact contents.

Reply:
Just to clarify, because I use the term "concrete artifact" a lot when I talk about inputs and outputs: the attributes modeled for the artifact inputs and outputs of a runnable are meant to describe _a range_ of potential artifacts, rather than a single unique artifact. In that sense, a runnable artifact input or output can be thought of as a class definition, whereas what I refer to as a "concrete artifact" is the instantiation of that class into an object: the actual contents (or the location) of a fully defined artifact that is used by the container at runtime. Another way to think about it is that a concrete artifact is an object that has a fully defined identity (e.g. a git revision, or a URL pointing to a concrete path in an S3 or GCS storage instance, or even a local filepath) that matches the description provided in the runnable input/output.

Concrete artifact values can be supplied as input parameters to a workflow run in many different ways:
* by pointing to an artifact tracked in one of the "stores" curated by FuseML. A gitea codeset is one example of this.
* by pointing to where the artifact is stored in an external storage service (e.g. an S3 URL, a GCS URL, or even a simple HTTP or FTP URL).
* when a runnable artifact input is connected to the output of another runnable, this can even be achieved locally, by passing the artifact contents as a pipeline workspace path/mount point

Reply:
Thanks for all that explanation. I think that now I have a better idea of it, and I agree that this information might be better in a proper documentation.